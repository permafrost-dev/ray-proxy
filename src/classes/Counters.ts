export class Counters {
    public totalBytesSent = 0;
    public totalRequestCounter = 0;
    public totalResponseTimeMs = 0;

    public constructor() {
        this.reset();
    }

    public reset(): this {
        this.totalBytesSent = 0;
        this.totalRequestCounter = 0;
        this.totalResponseTimeMs = 0;

        return this;
    }

    public onPayloadSent(payload: string): this {
        this.totalBytesSent += payload.length;

        return this;
    }

    public onRequestTime(requestTime: number): this {
        this.totalResponseTimeMs += requestTime;

        return this;
    }

    public onRequestSent(): this {
        this.totalRequestCounter += 1;

        return this;
    }
}
