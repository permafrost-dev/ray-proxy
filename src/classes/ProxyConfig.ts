import { existsSync } from 'fs';
import { findUpSync } from 'find-up';

export interface ProxyConfigurationData {
    hostName?: string;
    hostPort?: number;
    proxyPort?: number;
    displayBanner?: boolean;
}

export const defaultProxyConfigurationData = {
    hostName: 'localhost',
    hostPort: 23516,
    proxyPort: 23517,
    displayBanner: true,
};

export class ProxyConfig {
    public data: ProxyConfigurationData = defaultProxyConfigurationData;

    get hostName() {
        return this.data.hostName;
    }

    get hostPort() {
        return this.data.hostPort;
    }

    get proxyPort() {
        return this.data.proxyPort;
    }

    get displayBanner() {
        return this.data.displayBanner;
    }

    constructor(config: ProxyConfigurationData = {}) {
        this.useConfiguration(config);
    }

    public useConfiguration(config: ProxyConfigurationData): void {
        this.data = Object.assign({}, defaultProxyConfigurationData, config);
    }

    public static loadFromFile(filename: string | null = null): ProxyConfig {
        if (!filename) {
            filename = findUpSync('ray-proxy.config.js') || null;

            if (filename === undefined) {
                filename = __dirname + '/ray-proxy.config.js';
            }
        }

        if (filename === null || !existsSync(filename)) {
            //throw new Error('config file not found.');
            console.log('config file not found, using defaults');

            return new ProxyConfig(defaultProxyConfigurationData);
        }

        let configData: ProxyConfigurationData;

        try {
            configData = require(filename); //<ProxyConfigurationData>JSON.parse(contents);
            return new ProxyConfig(configData);
        } catch (err) {
            console.log('Could not load config file:', err);

            throw err;
        }
    }
}
