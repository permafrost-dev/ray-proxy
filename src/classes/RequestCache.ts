export class RequestCache {
    public countCache: number[] = [];
    public times: Record<number, number> = {};

    /**
     * add an entry to the request history.
     *   you should call cache() once on each request
     *
     * @param currentTs number|null
     */
    public cache(requestId: any, currentTs: number | null = null, purgeOldItems = true): void {
        this.times[requestId] = currentTs ?? new Date().getTime();

        currentTs = Number(((currentTs ?? new Date().getTime()) / 1000).toFixed(0));

        this.countCache.push(currentTs);

        if (this.countCache.length > 3000) {
            this.countCache.splice(0, this.countCache.length - 3000);
        }

        if (purgeOldItems) {
            this.purgeCache(currentTs);
        }
    }

    /**
     * returns the number of requests per second based on the current request cache history.
     *
     * @returns number|float
     */
    public requestsPerSec(): number {
        if (!this.countCache.length) {
            return 0.0;
        }

        const firstTs = this.countCache[0];
        const lastTs = this.countCache[this.countCache.length - 1];
        let elapsedTime = lastTs - firstTs;

        if (elapsedTime <= 0) {
            elapsedTime = 1;
        }

        // console.log({ firstTs, lastTs, length: this.countCache.length, elapsedTime });

        return this.countCache.length / elapsedTime;
    }

    /**
     *
     * @param currentTs number
     * @param maxAgeInSeconds number
     */
    public purgeCache(currentTs: number, maxAgeInSeconds = 5): void {
        // purge entries older than maxAgeInSeconds sec
        this.countCache = this.countCache.filter(ts => currentTs - ts <= maxAgeInSeconds);
    }
}
