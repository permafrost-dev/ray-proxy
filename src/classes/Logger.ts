import { Counters } from '@/classes/Counters';
import { RequestCache } from '@/classes/RequestCache';
import { formatPayloadSize } from '@/lib/utils';

export class Logger {
    public counters: Counters;
    public requestCache: RequestCache;
    public logger: any;

    constructor(counters: Counters, requestCache: RequestCache, logger: any = null) {
        this.logger = logger ?? console;
        this.counters = counters;
        this.requestCache = requestCache;
    }

    public log(...args: any[]) {
        this.logger.log(...args);
    }

    public separator() {
        this.log('---');
    }

    public logRayAppStats(requestTime: number) {
        const avgTime = (this.counters.totalResponseTimeMs / this.counters.totalRequestCounter).toFixed(1);

        this.log(`[Ray App] response time from Ray (last): ${requestTime} ms`);
        this.log(`[Ray App] response time from Ray (avg): ${avgTime} ms`);
    }

    public logProjectStats() {
        this.log(`[project] requests/sec to Ray app: ${this.requestCache.requestsPerSec().toFixed(2)}`);
        this.log(`[payload] data sent (total): ${formatPayloadSize(this.counters.totalBytesSent)} kb`);
    }

    public logPayloadStats(payload: string) {
        const avgRequestBytes = this.counters.totalBytesSent / this.counters.totalRequestCounter;

        this.log(`[payload] data sent (last payload): ${formatPayloadSize(payload.length)} kb`);
        this.log(`[payload] data sent (avg/request): ${formatPayloadSize(avgRequestBytes)} kb`);
    }

    public logPayload(payload: string) {
        this.log('payload: ', payload);
    }

    public display(requestTime: number, payload: string) {
        this.separator();
        this.logPayload(payload);
        this.separator();
        this.logRayAppStats(requestTime);
        this.logProjectStats();
        this.logPayloadStats(payload);
        this.separator();
    }
}
