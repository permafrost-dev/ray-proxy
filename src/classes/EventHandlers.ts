/* eslint-disable no-unused-vars */

import { relayResponseFromAppToClient, sendPreflightCorsResponse } from '../lib/utils';
import { Counters } from './Counters';
import { Logger } from './Logger';
import { ProxyConfig } from './ProxyConfig';
import { RequestCache } from './RequestCache';

export class EventHandlers {
    onHead(config: ProxyConfig, axios: any) {
        return async (req: any, reply: any) => {
            let response: any;

            try {
                response = await axios.head(`http://${config.hostName}:${config.hostPort}`);
            } catch (err) {
                response = err.response;
            }

            relayResponseFromAppToClient(reply, response);
        };
    }

    onGet(config: ProxyConfig, axios: any) {
        return async (req: any, reply: any) => {
            let response: any;

            try {
                response = await axios.get(`${req.url}`);
            } catch (err) {
                response = err.response;
            }

            relayResponseFromAppToClient(reply, response);
        };
    }

    onOptions(config: ProxyConfig, axios: any) {
        return async (req: any, reply: any) => {
            //const response = await axios.options(`http://${config.hostName}:${config.hostPort}`);
            //relayResponseFromAppToClient(reply, response);
            sendPreflightCorsResponse(reply);
        };
    }

    onPost(config: ProxyConfig, axios: any, requestCache: RequestCache, counters: Counters, logger: Logger) {
        return async (request: any, reply: any) => {
            requestCache.cache(request.id, new Date().getTime());

            const payload = JSON.stringify(request.body).toString();

            const response = await axios.post(`http://${config.hostName}:${config.hostPort}`, request.body);

            relayResponseFromAppToClient(reply, response);

            const requestTime = new Date().getTime() - (requestCache.times[request.id] ?? 0);

            counters.onRequestSent().onPayloadSent(payload).onRequestTime(requestTime);
            logger.display(requestTime, payload);
        };
    }
}
