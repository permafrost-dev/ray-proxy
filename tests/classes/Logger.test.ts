/* eslint-disable no-undef */

import { Counters } from '../../src/classes/Counters';
import { Logger } from '../../src/classes/Logger';
import { RequestCache } from '../../src/classes/RequestCache';

class FakeConsole {
    public loggedData: any[] = [];

    public log(...args: any[]) {
        this.loggedData.push(...args);
    }
}

it('displays the statistics', () => {
    const counters = new Counters();
    counters.totalBytesSent = 1024;
    counters.totalRequestCounter = 1;
    counters.totalResponseTimeMs = 100;

    const cache = new RequestCache();
    cache.cache('req-1', 100000000, false);

    const fakeConsole = new FakeConsole();

    const logger = new Logger(counters, cache, fakeConsole);

    logger.display(100, 'test payload');

    expect(fakeConsole.loggedData).toMatchSnapshot();
});
