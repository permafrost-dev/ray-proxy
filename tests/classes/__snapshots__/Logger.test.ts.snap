// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays the statistics 1`] = `
Array [
  "---",
  "payload: ",
  "test payload",
  "---",
  "[Ray App] response time from Ray (last): 100 ms",
  "[Ray App] response time from Ray (avg): 100.0 ms",
  "[project] requests/sec to Ray app: 1.00",
  "[payload] data sent (total): 1 kb",
  "[payload] data sent (last payload): 0.012 kb",
  "[payload] data sent (avg/request): 1 kb",
  "---",
]
`;
